### PR TITLE
Validate sandbox not mapping

### PR DESCRIPTION
**Pull request by xichan-ms**
[AB#345644](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/345644)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/6896)

### CODE REVIEW

File: src/docfx/build/learn/LearnHierarchyBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        {
+            _contentValidator.ValidateHierarchy(GetAllLearnHierarchyModels().ToList());
+        }
+
+        public void AddLearningPath(JObject content)
+        {
+            var path = JsonUtility.ToObject<LearningPath>(_errors, content);
+            _learningPaths.Add(path);
+
+            if (path.Trophy != null && GetAchievement(path.Trophy, out var achievement) && achievement != null)
```
Comment (yufeih) [inline src/docfx/build/learn/LearnHierarchyBuilder.cs:38]: Ensure this method is idempotent, that is, the method should merely add the item to the collection. Move these achievement logic to the `ValidateHierarchy` method.

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 2]
```diff
@@ -80,7 +80,7 @@ public ExternalXref[] GetValidateExternalXrefs()
             }).OrderBy(externalXref => externalXref.Uid).ToArray();
         }
 
-        public JToken TransformContent(ErrorBuilder errors, FilePath file)
+        public (JToken, JToken) TransformContent(ErrorBuilder errors, FilePath file)
```
Comment (yufeih) [inline src/docfx/lib/schema/JsonSchemaTransformer.cs:83]: The models should be de-serialized from transformed models, if there are source info problem, we could fix them later by propagating source info into the de-serialized JToken.
```suggestion
        public JToken TransformContent(ErrorBuilder errors, FilePath file)
```

File: src/docfx/build/page/PageBuilder.cs [hunk 3]
```diff
@@ -286,15 +289,38 @@ private JObject LoadSchemaDocument(ErrorBuilder errors, FilePath file)
...
+            var (content, token) = _jsonSchemaTransformer.TransformContent(errors, file);
             if (!(content is JObject transformedContent))
             {
                 throw Errors.JsonSchema.UnexpectedType(new SourceInfo(file, 1, 1), JTokenType.Object, content.Type).ToException();
             }
+            else if (token is JObject fileContent)
+            {
+                switch (mime.Value?.ToLowerInvariant())
+                {
+                    case Constants.LearningPath:
```
Comment (yufeih) [inline src/docfx/build/page/PageBuilder.cs:304]: We don't use `Constants` in docfx, if this string needs sharing, it probably indicates some kind of code smell.
```suggestion
                    case "learningpath":
```

File: src/docfx/lib/learn/Module.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace Microsoft.Docs.Build
+{
+    internal class Module
+    {
+        public SourceInfo<string> Uid { get; set; }
+
+        public SourceInfo<string>[]? Units { get; set; }
+
+        public Trophy? Badge { get; set; }
```
Comment (yufeih) [inline src/docfx/lib/learn/Module.cs:12]:
```suggestion
        public Badge? Badge { get; set; }
```

File: src/docfx/lib/learn/Achievement.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Docs.Build
+{
+    internal class Achievement
+    {
+        public SourceInfo<string> Uid { get; set; }
+
+        public string Type { get; set; } = "";
```
Comment (yufeih) [inline src/docfx/lib/learn/Achievement.cs:10]: Use enum over strings.
```suggestion
        public AchievementType Type { get; set; };
```

File: src/docfx/build/learn/LearnHierarchyBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                _achievements.Add(achievement);
+            }
+        }
+
+        public void AddModule(JObject content)
+        {
+            var module = JsonUtility.ToObject<Module>(_errors, content);
+            _modules.Add(module);
+
+            if (module.Badge != null && GetAchievement(module.Badge, out var achievement) && achievement != null)
```
Comment (yufeih) [inline src/docfx/build/learn/LearnHierarchyBuilder.cs:49]: The type is fixed, can simply `Add(new Achievement { Type = Badge })`.

File: azure-pipelines-v3.yml [hunk 7]
```diff
@@ -213,6 +213,8 @@ jobs:
         params: https://github.com/MicrosoftDocs/bot-docs-pr --timeout 40 --no-dry-sync --dry-run --public-template
       learn-pr:
         params: https://github.com/MicrosoftDocs/learn-pr --timeout 100 --no-dry-sync
+      learn-pr-regression:
```
[the remark this answers is not in the document]
Comment (xichan-ms) [reply azure-pipelines-v3.yml]: OK

File: azure-pipelines-v3.yml [hunk 8]
```diff
@@ -213,6 +213,8 @@ jobs:
         params: https://github.com/MicrosoftDocs/bot-docs-pr --timeout 40 --no-dry-sync --dry-run --public-template
       learn-pr:
         params: https://github.com/MicrosoftDocs/learn-pr --timeout 100 --no-dry-sync
+      regression.learn-pr:
```
[the remark this answers is not in the document]
Comment (xichan-ms) [reply azure-pipelines-v3.yml:216]: OK

File: src/docfx/lib/learn/Achievement.cs [hunk 9]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) Microsoft. All rights reserved.
```
[the remark this answers is not in the document]
Comment (xichan-ms) [reply src/docfx/lib/learn/Achievement.cs:1]: OK

File: src/docfx/lib/learn/Achievement.cs [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Docs.Build
+{
+    internal class Achievement
+    {
+        public SourceInfo<string> Uid { get; set; }
```
[the remark this answers is not in the document]
Comment (xichan-ms) [reply src/docfx/lib/learn/Achievement.cs:8]: OK

File: src/docfx/build/learn/LearnHierarchyBuilder.cs [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+            _errors = errors;
+            _contentValidator = contentValidator;
+        }
+
+        public void ValidateHierarchy()
+        {
+            _contentValidator.ValidateHierarchy(GetAllLearnHierarchyModels().ToList());
+        }
+
+        public void AddLearningPath(JObject content)
```
[the remark this answers is not in the document]
Comment (xichan-ms) [reply src/docfx/build/learn/LearnHierarchyBuilder.cs:33]: OK

File: src/docfx/lib/schema/JsonSchemaTransformer.cs (hunk 2 above)
Comment (xichan-ms) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs:83]: OK

File: src/docfx/build/page/PageBuilder.cs (hunk 3 above)
Comment (xichan-ms) [reply src/docfx/build/page/PageBuilder.cs:304]: OK

File: src/docfx/lib/learn/Module.cs (hunk 4 above)
Comment (xichan-ms) [reply src/docfx/lib/learn/Module.cs:12]: OK

File: src/docfx/lib/learn/Achievement.cs (hunk 5 above)
Comment (xichan-ms) [reply src/docfx/lib/learn/Achievement.cs:10]: OK

File: src/docfx/build/learn/LearnHierarchyBuilder.cs (hunk 6 above)
Comment (xichan-ms) [reply src/docfx/build/learn/LearnHierarchyBuilder.cs:49]: ok

File: src/docfx/build/learn/LearnHierarchyBuilder.cs (hunk 1 above)
Comment (xichan-ms) [reply src/docfx/build/learn/LearnHierarchyBuilder.cs:38]: OK

File: docs/specs/validation/hierarchy-validation.yml [hunk 12]
```diff
@@ -0,0 +1,117 @@
+# Hierarchy validation - learn, module not sandbox mapping
+inputs:
+  docfx.yml: |
+    markdownValidationRules: rules.json
+    sandboxEnabledModuleList: mapping.json
```
[the remark this answers is not in the document]
Comment (yufeih) [reply docs/specs/validation/hierarchy-validation.yml:5]: We could consolidate them later.

File: src/docfx/learn/Achievement.cs [hunk 13]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Docs.Build
+{
+    internal class Achievement
+    {
+        public SourceInfo<string> Uid { get; set; } = new SourceInfo<string>("");
```
Comment (yufeih) [inline src/docfx/learn/Achievement.cs:8]: prefer _init only properties_
```suggestion
        public SourceInfo<string> Uid { get; init; } = new SourceInfo<string>("");
```

**Review by yufeih — APPROVED** [review]

File: src/docfx/learn/Achievement.cs (hunk 13 above)
Comment (xichan-ms) [reply src/docfx/learn/Achievement.cs]: OK

**Comment by yufeih** [comment]
nice job spell checker bot 🤣

**Comment by xichan-ms** [comment]
Diff explanations:
The diff is duo to the ruleset changing from PROD to RegressionRuleset and the content is expected.

**Comment by OsmondJiang** [comment]
👍👍

**Pull request merged**